OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsLogic.cs

```diff
@@ -0,0 +1,160 @@
+		readonly List<Tuple<string, string>> panels = new List<Tuple<string, string>>();
+		string activePanel = "";
+
+		public static bool NeedsRestart = false;
+
+		static SettingsLogic() { }
+
+		[ObjectCreator.UseCtor]
+		public SettingsLogic(Widget widget, Action onExit, WorldRenderer worldRenderer, Dictionary<string, MiniYaml> logicArgs)
+		{
+			panelContainer = widget.Get("PANEL_CONTAINER");
+			var panelTemplate = panelContainer.Get<ContainerWidget>("PANEL_TEMPLATE");
+			panelContainer.RemoveChild(panelTemplate);
+
+			tabContainer = widget.Get("TAB_CONTAINER");
+			tabTemplate = tabContainer.Get<ButtonWidget>("BUTTON_TEMPLATE");
+			tabContainer.RemoveChild(tabTemplate);
+
+			if (logicArgs.TryGetValue("ButtonStride", out var buttonStrideNode))
+				buttonStride = FieldLoader.GetValue<int2>("ButtonStride", buttonStrideNode.Value);
+
+			if (logicArgs.TryGetValue("Panels", out var settingsPanels))
+			{
+				foreach (var settingsPanel in settingsPanels.Nodes)
+				{
+					var panelID = settingsPanel.Value.Nodes.FirstOrDefault(n => n.Key == "Id").Value.Value;
+					var label = settingsPanel.Value.Nodes.FirstOrDefault(n => n.Key == "Label").Value.Value;
+					var container = panelTemplate.Clone() as ContainerWidget;
+					container.Id = panelID;
+					panelContainer.AddChild(container);
+
+					panels.Add(Tuple.Create(panelID, label));
+
+					Game.LoadWidget(worldRenderer.World, panelID, container, new WidgetArgs()
+					{
+						{ "settingsLogic", this },
+						{ "panelID", panelID },
+						{ "label", label }
+					});
+				}
+			}
+
+			widget.Get<ButtonWidget>("BACK_BUTTON").OnClick = () =>
+			{
+				leavePanelActions[activePanel]();
+				var current = Game.Settings;
+				current.Save();
+
+				Action closeAndExit = () => { Ui.CloseWindow(); onExit(); };
+				if (NeedsRestart)
+				{
+					Action restart = () =>
+					{
+						var external = Game.ExternalMods[ExternalMod.MakeKey(Game.ModData.Manifest)];
+						Game.SwitchToExternalMod(external, null, closeAndExit);
+					};
+
+					ConfirmationDialogs.ButtonPrompt(
+						title: "Restart Now?",
+						text: "Some changes will not be applied until\nthe game is restarted. Restart now?",
+						onConfirm: restart,
+						onCancel: closeAndExit,
+						confirmText: "Restart Now",
+						cancelText: "Restart Later");
+				}
+				else
+					closeAndExit();
+			};
+
+			widget.Get<ButtonWidget>("RESET_BUTTON").OnClick = () =>
+			{
+				Action reset = () =>
+				{
+					resetPanelActions[activePanel]();
+					Game.Settings.Save();
+				};
+
+				ConfirmationDialogs.ButtonPrompt(
+					title: "Reset \"{0}\"".F(panels.Find(p => p.Item1 == activePanel).Item2),
+					text: "Are you sure you want to reset\nall settings in this panel?",
+					onConfirm: reset,
+					onCancel: () => { },
+					confirmText: "Reset",
+					cancelText: "Cancel");
+			};
+		}
+
+		public void RegisterSettingsPanel(string panelID, string label, Func<Widget, Action> init, Func<Widget, Action> reset)
```

Comment:
This method keeps the same logic as before only now each separate chrome logic class registers itself.